File: 18_자율문제/6월2주차/박지원/pgisland.js
//무인도 여행 자바스크립트
//자바만큼 익숙해지게 많이 풀어보기
//이 문제는 자바로 푸는거랑 푸는 방식이 아예 똑같았다. 방문처리, x가 아닌 곳을 찾아 bfs 를 돌리고 answer처리해주면 됨

function solution(maps) {
    var answer = [];
    const [x, y] = [maps.length, maps[0].length];
    const visited = Array.from({length: x}, () => Array(y).fill(0));
    const move = [[0, 1], [1, 0], [-1, 0], [0, -1]];
    
    const bfs = (a, b) => {
        let cnt = 0;
        const q = [[a, b]];
        cnt += parseInt(maps[a][b]);
        visited[a][b] = 1;
        while(q.length) {
            const [r, c] = q.shift();
            for (let i = 0; i < 4; i++) {
                const ma = r + move[i][0];
                const mb = c + move[i][1];
                if (ma >= 0 && mb >= 0 && ma < x && mb < y && !visited[ma][mb] && maps[ma][mb] != 'X') {
                    visited[ma][mb] = 1;
                    cnt += parseInt(maps[ma][mb]);
                    q.push([ma, mb]);
                }
                
            }
        }
        answer.push(cnt);
    }
    
    for (let i = 0; i < x; i++) {
        for (let j = 0; j < y; j++) {
            if (!visited[i][j] && maps[i][j] != 'X') {
                bfs(i, j);
            }
        }
    }
    
    if (answer.length == 0) return [-1]
    return answer.sort((a, b) => a - b);
}

File: 18_자율문제/6월2주차/박지원/pgisland.java
//무인도 여행 자바로 먼저 풀기
//코드가.. 좀.. 바보같은 코드같다.. stream을 잘써야할듯..

import java.util.*;

class Node {
    int x;
    int y;
    public Node(int x, int y) {
        this.x = x;
        this.y = y;
    }
}
class Solution {
    static int[] dx = {-1, 0, 1, 0};
    static int[] dy = {0, 1, 0, -1};
    static char[][] map;
    static int n, m;
    static boolean[][] visited;
    static ArrayList<Integer> answer;
    public int[] solution(String[] maps) {
        n = maps.length;
        m = maps[0].length();
        answer = new ArrayList<>();
        visited = new boolean[n][m];
        map = new char[n][m];
        for (int i = 0; i < n; i++) {
            String s = maps[i];
            for (int j = 0; j < m; j++) {
                map[i][j] = s.charAt(j);
            }
        }
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (map[i][j] != 'X' && !visited[i][j]) {
                    bfs(i, j);
                }
            }
        }
        int[] ans;
        if (answer.size() <= 0) {
            ans = new int[1];
            ans[0] = -1;
        } else {
            ans = new int[answer.size()];
            for (int i = 0; i < answer.size(); i++) {
                ans[i] = answer.get(i);
            }
          Arrays.sort(ans);  
        } 
        return ans;
    }
    static void bfs(int x, int y) {
        int count = 0;
        Queue<Node> q = new LinkedList<>();
        visited[x][y] = true;
        count += map[x][y] - '0';
        q.add(new Node(x, y));
        while (!q.isEmpty()) {
            Node cur = q.poll();
            for (int i = 0; i < 4; i++) {
                int mx = cur.x + dx[i];
                int my = cur.y + dy[i];
                if (mx < n && my < m && mx >= 0 && my >= 0 && !visited[mx][my] && map[mx][my] != 'X') {
                    visited[mx][my] = true;
                    count += map[mx][my] - '0';
                    q.add(new Node(mx, my));
                }
            }
        }
        answer.add(count);
    }
}
